### ISSUE TITLE
Weight_decay handling should be done in `backwarding`

### ISSUE THREAD

**Issue by zhoonit** [nnstreamer/nntrainer#213]
When weight decay is added to a layer, `Optimizer::calculate` takes `weight_decay` term and use that to update `djdw`

I think it should be done in the `layer::backwarding` phase and `optimizer` shouldn't care about `weight_decay` term as it is not about optimizing but updating derivatives.

for example `FullyConnectedLayer::backwarding` should be like

```cpp
FullyConnectedLayer::backwarding(Tensor derivative, int iteration) {
   if(should_apply_weight_decay) {
      djdw.add_i(weight, weight_decay.lambda);
   }

   /** rest of the calculation */
}
```

**Comment by kparichay** [comment]
The above works as long as the decaying of the learning rate is not applied to weight decay, which I think is.
@zhoonit as our benchmark is tf/tflite, can you please verify with it if they applying learning rate decay to the constant of weight decay?

**Comment by zhoonit** [comment]
I looked through the implementation and specs of `tf.keras.optimizers`.
Weight_decay is not considered in the optimizer nor optimizer accepts weight_decay term.

Some of the advanced optimizer(residing in `tfa.optimizers`) like `AdamW` does accept weight decay term because they treat weight decay differently (not updating). But those kind of optimizer does not update `dw` AKAIK.

**Comment by jijoongmoon** [comment]
By the way, I think current implementation is something wrong. Looks like weight_decay is applied twice.

**Comment by zhoonit** [comment]
> By the way, I think current implementation is something wrong. Looks like weight_decay is applied twice.

Yes. It is applied twice. I'll work on that with fixing the issue :)
